tests/engine/processes/test_builder.py

```diff
@@ -8,49 +8,442 @@
+
+
+def test_merge():
+    """Test the ``_merge`` method to merge a dictionary into an existing builder."""
+    builder = NestedNamespaceProcess.get_builder()
+    builder.nested.namespace = {'int': 1, 'float': 2.0}
+    assert builder._inputs(prune=True) == {'nested': {'namespace': {'int': 1, 'float': 2.0}}}
+
+    # Define only one of the required ports of `nested.namespace`. This should leave the `float` input untouched and
+    # even though not specified explicitly again, since the merged dictionary still contains it, the
+    # `nested.namespace` port should still be valid.
+    builder._merge({'nested': {'namespace': {'int': 5}}})
+    assert builder._inputs(prune=True) == {'nested': {'namespace': {'int': 5, 'float': 2.0}}}
+
+    # Perform same test but passing the dictionary in as keyword arguments
+    builder._merge(**{'nested': {'namespace': {'int': 5}}})
+    assert builder._inputs(prune=True) == {'nested': {'namespace': {'int': 5, 'float': 2.0}}}
+
+
+def test_instance_interference():
+    """Test that two builder instances do not interact through the class.
+
+    This is a regression test for #4420.
+    """
+
+    class ProcessOne(Process):
+        """Process with nested required ports to check the update functionality."""
+
+        @classmethod
+        def define(cls, spec):
+            super().define(spec)
+            spec.input('port', valid_type=int, default=1)
+
+    class ProcessTwo(Process):
+        """Process with nested required ports to check the update functionality."""
+
+        @classmethod
+        def define(cls, spec):
+            super().define(spec)
+            spec.input('port', valid_type=int, default=2)
+
+    builder_one = ProcessOne.get_builder()
+    assert builder_one.port == 1
+    assert isinstance(builder_one, ProcessBuilderNamespace)
+
+    # Create a second builder and check its only port has the correct default, but also that the original builder
+    # still has its original port default and hasn't been affected by the second builder because the share a port
+    # name.
+    builder_two = ProcessTwo.get_builder()
+    assert isinstance(builder_two, ProcessBuilderNamespace)
+    assert builder_two.port == 2
+    assert builder_one.port == 1
+
+
+def test_pretty_repr(example_inputs):
+    """Test the pretty representation of the ``ProcessBuilder`` class."""
+    builder = ExampleWorkChain.get_builder()
+    builder._update(example_inputs)
+
+    pretty_repr = \
+    """Process class: ExampleWorkChain
+    Inputs:
+    {
+        "metadata": {},
+        "dynamic": {
+            "namespace": {
+                "alp": 1
+            }
+        },
+        "name": {
+            "spaced": 1
+        },
+        "name_spaced": "underscored",
+        "dict": {
+            "a": 1,
+            "b": {
+                "c": 2
+            }
+        },
+        "boolean": true
+    }"""
+    pretty_repr = re.sub(r'(?m)^\s{4}', '', pretty_repr)
```

Comment:
Instead of this, I think it is better to use the `dedent` method from the `textwrap` package. This is clearer I think to what it is doing and we already use it elsewhere in the tests. See comment above how to do it
```suggestion
    pretty_repr = textwrap.dedent(
    """Process class: ExampleWorkChain
    Inputs:
    {
        "metadata": {},
        "dynamic": {
            "namespace": {
                "alp": 1
            }
        },
        "name": {
            "spaced": 1
        },
        "name_spaced": "underscored",
        "dict": {
            "a": 1,
            "b": {
                "c": 2
            }
        },
        "boolean": true
    }""")
```